lib/memoryprofiler/src/trait.rs

```diff
@@ -0,0 +1,169 @@
+pub trait MemoryUsage {
+    fn size_of(&self) -> usize;
+}
+
+/// Primitive types
+impl MemoryUsage for bool {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<bool>()
+    }
+}
+impl MemoryUsage for char {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<char>()
+    }
+}
+impl MemoryUsage for f32 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<f32>()
+    }
+}
+impl MemoryUsage for f64 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<f64>()
+    }
+}
+impl MemoryUsage for i8 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i8>()
+    }
+}
+impl MemoryUsage for i16 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i16>()
+    }
+}
+impl MemoryUsage for i32 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i32>()
+    }
+}
+impl MemoryUsage for i64 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i64>()
+    }
+}
+impl MemoryUsage for isize {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<isize>()
+    }
+}
+impl MemoryUsage for u8 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u8>()
+    }
+}
+impl MemoryUsage for u16 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u16>()
+    }
+}
+impl MemoryUsage for u32 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u32>()
+    }
+}
+impl MemoryUsage for u64 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u64>()
+    }
+}
+impl MemoryUsage for usize {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<usize>()
+    }
+}
+
+// pointer
+// Pointers aren't necessarily safe to dereference, even if they're nonnull.
+
+// references
+impl<T: MemoryUsage> MemoryUsage for &T {
+    fn size_of(&self) -> usize {
+        (**self).size_of()
+    }
+}
+impl<T: MemoryUsage> MemoryUsage for &mut T {
+    fn size_of(&self) -> usize {
+        (**self).size_of()
+    }
+}
+
+// slices
+impl<T: MemoryUsage> MemoryUsage for [T] {
+    fn size_of(&self) -> usize {
+        self.iter().map(MemoryUsage::size_of).sum()
+    }
+}
+
+// arrays
+impl<T: MemoryUsage, const N: usize> MemoryUsage for [T; N] {
+    fn size_of(&self) -> usize {
+        self.iter().map(MemoryUsage::size_of).sum()
+    }
+}
+
+// strs
+impl MemoryUsage for str {
+    fn size_of(&self) -> usize {
+        self.as_bytes().size_of()
```

Comment:
Hmm okay this is interesting, will need testing to make sure we don't over or under count but I think `str` is the same as `[u8]`; meaning I'm not sure what it's overhead is: `&str` and `&[u8]` should have an extra 2 `usize` on the stack though!